File: 2. Array and String/reverse_words/reverse_words.js
/**
 * @param {string} s
 * @return {string}
 */
//  var reverseWords = function(s) {
//     let words = s.split(' ');
//     for(let index=0; index<words.length; index++){
//         let wordArr = words[index].split('');
//         words[index] = wordArr.reverse().join('');
//     }
//     return words.join(' ');
// };

var reverseWords = function(s) {
    let finalArr = [];
    let j = 0;
    for(let i=s.length-1; i>=0; i--){
        if(s[i] === ' '){
            j++;
            continue;
        }
        if(!finalArr[j]) finalArr[j] = '';
        finalArr[j] = finalArr[j] + s[i];
    }
    return finalArr.reverse().join(' ');
}

// console.log(reverseWords("Let's take LeetCode contest"));

File: 2. Array and String/11. reverse_words/reverse_words.py
from collections import deque
class Solution:
    def reverseWords(self, s: str) -> str:
        wordList = deque()
        charList = []
        for i in range(len(s)-1,-1,-1):
            if(s[i] == " "):
                wordList.appendleft("".join(charList))
                charList = []
            charList.append(s[i])

        wordList.appendleft("".join(charList))
        print(wordList)
        return " ".join(wordList)


    def reverseWords1(self, s: str) -> str:
	    return " ".join([word[::-1] for word in s.split()])

    # Using 2 pointers - Leetcode solution
    def reverseWords2(self, s: str) -> str:
        startIndex = 0
        endIndex = 0
        wordIndex = 0
        s = list(s)
        while startIndex <= (len(s)-1):
            while(wordIndex <= (len(s)-1) and s[wordIndex] != " "):
                wordIndex += 1
            endIndex = wordIndex - 1
            
            while startIndex < endIndex:
                s[startIndex], s[endIndex] = s[endIndex], s[startIndex]
                startIndex += 1
                endIndex -= 1
            
            wordIndex += 1
            startIndex = wordIndex
            endIndex = startIndex
        s = "".join(s)
        return s


test = Solution()
print(test.reverseWords("Let's take LeetCode contest"))
print(test.reverseWords("God Ding"))
print(test.reverseWords1("Let's take LeetCode contest"))
print(test.reverseWords1("God Ding"))
print(test.reverseWords2("Let's take LeetCode contest"))
print(test.reverseWords2("God Ding"))